### ISSUE TITLE
MatchingFields fails when used as option in DeleteAllOf(), "field label not supported"

### ISSUE THREAD

**Issue by albert-zhong** [kubernetes-sigs/controller-runtime#1956]
I have a CRD called `Folder`, which contains the field `spec.folderName`. I created an index on it:
```go
	if err := mgr.GetFieldIndexer().IndexField(
		context.Background(),
		&fv1.Folder{},
	        "spec.folderName",
		func(obj client.Object) []string {
			tfc := obj.(*fv1.Folder)
			return []string{tfc.Spec.FolderName}
		},
	); err != nil {
		return err
	}
```

I have a function that deletes all Folder CRs with a particular value of `spec.folderName` via the MatchingFields option:
```go
        func (r *FolderReconciler) deleteFolders(ctx context.Context, req ctrl.Request, folderName string) (ctrl.Result, error) {
	  if err := r.Client.DeleteAllOf(ctx,
		  &fv1.Folder{},
		  client.InNamespace(req.Namespace),
		  client.MatchingFields{"spec.folderName": folderName},
	  ); err != nil {
		  return ctrl.Result{}, err
	  }
        }
```
However, DeleteAllOf() always fails due to "field label not supported: spec.folderName."
```
2022-07-14T13:40:39.425-0700	ERROR	controller.folder	Reconciler error	{"reconciler group": "test.folder-k8s.io", "reconciler kind": "Folder", "name": "f1", "namespace": "default", "error": "field label not supported: spec.folderName"}
```
I expect DeleteAllOf() to delete all Folder CRs with the given value spec.folderName. This issue is confusing me because I can call `List()` using the same `MatchingFields `option and `Delete()` each folder individually without issue. Has anyone experienced this issue?

**Comment by FillZpp** [comment]
The `IndexField` you created is only registered in local Informer for you listing from cache. However, APIServer can not recognize it when you submit a DeleteAllOf request to it.

**Comment by albert-zhong** [comment]
> The `IndexField` you created is only registered in local Informer for you listing from cache. However, APIServer can not recognize it when you submit a DeleteAllOf request to it.

Thanks! Is there a way to get APIServer to recognize the IndexField? Or is trying to use IndexFields with `DeleteAllOf()` completely impossible?

**Comment by FillZpp** [comment]
No, it's impossible. Unless you also put the field into labels, so you can use label selector to delete them.

**Comment by albert-zhong** [comment]
Thank you for the insight!
